File: javascript/queen-attack/queen-attack.js
export class QueenAttack {
    constructor(position = {white: [0,3], black: [7,3]}) {
        this.white = position.white
        this.black = position.black

        if(this.white.every((v,i) => v === this.black[i])){
            throw new Error('Queens cannot share the same space')
        }
        
        this.posW = this.white[0] * 8 + this.white[1] 
        this.posB = this.black[0] * 8 + this.black[1]
    }
    
    toString() {
        let board = ''
        for(let i = 0; i < 64; i++) {
            if (this.posW == i)
                board += "W"
            else if (this.posB == i)
                board += "B"
            else
                board += '_'
            board += (i + 1) % 8 == 0 ? "\n" : " "
        }
        return board
    } 

    canAttack() {
        let diff = Math.abs(this.posW - this.posB)
        //horizontally
        if (this.white[0] == this.black[0])
            return true
        //vertically
        if (this.white[1] == this.black[1])
            return true
        //diagonally
        for (let i = 0; i < 8; i++){
            if(diff == (8 * i + i) || diff == (8 * i - i)) 
                return true
        }
        
        return false
    }
}

File: python/queen-attack/queen_attack.py
class Queen:
    def __init__(self, row, column):
        if row < 0:
            raise ValueError("row not positive")
        if column < 0:
            raise ValueError("column not positive")
        if row > 7:
            raise ValueError("row not on board")
        if column > 7:
            raise ValueError("column not on board")

        self.row, self.column = row, column

    def can_attack(self, another_queen):
        q1, q2 = self, another_queen

        if q1.row == q2.row and q1.column == q2.column:
            raise ValueError("Invalid queen position: both queens in the same square")

        if q1.row == q2.row or q1.column == q2.column:
            return True
        elif abs(q1.row - q2.row) == abs(q1.column - q2.column):
            return True

        return False
